packages/wrangler/src/api/startDevWorker/ProxyController.ts

```diff
@@ -0,0 +1,520 @@
+			}
+
+			websocket.send(JSON.stringify(message));
+		} catch (cause) {
+			const error = castErrorCause(cause);
+
+			if (retries > 0) {
+				return this.sendMessageToInspectorProxyWorker(message, retries - 1);
+			}
+
+			this.emitErrorEvent(
+				`Failed to send message to InspectorProxyWorker: ${JSON.stringify(
+					message
+				)}`,
+				error
+			);
+
+			throw error;
+		}
+	}
+
+	// ******************
+	//   Event Handlers
+	// ******************
+
+	onConfigUpdate(data: ConfigUpdateEvent) {
+		this.latestConfig = data.config;
+		this.createProxyWorker();
+
+		void this.sendMessageToProxyWorker({ type: "pause" });
+	}
+	onBundleStart(data: BundleStartEvent) {
+		this.latestConfig = data.config;
+
+		void this.sendMessageToProxyWorker({ type: "pause" });
+	}
+	onReloadStart(data: ReloadStartEvent) {
+		this.latestConfig = data.config;
+
+		void this.sendMessageToProxyWorker({ type: "pause" });
+	}
+	onReloadComplete(data: ReloadCompleteEvent) {
+		this.latestConfig = data.config;
+		this.latestBundle = data.bundle;
+
+		void this.sendMessageToProxyWorker({
+			type: "play",
+			proxyData: data.proxyData,
+		});
+
+		void this.sendMessageToInspectorProxyWorker({
+			type: "reloadComplete",
+			proxyData: data.proxyData,
+		});
+	}
+	onProxyWorkerMessage(message: ProxyWorkerOutgoingRequestBody) {
+		switch (message.type) {
+			case "previewTokenExpired":
+				this.emitPreviewTokenExpiredEvent(message.proxyData);
+
+				break;
+			case "error":
+				this.emitErrorEvent("Error inside ProxyWorker", message.error);
+
+				break;
+			case "debug-log":
+				logger.debug("[ProxyWorker]", ...message.args);
+
+				break;
+			default:
+				assertNever(message);
+		}
+	}
+	onInspectorProxyWorkerMessage(
+		message: InspectorProxyWorkerOutgoingWebsocketMessage
+	) {
+		switch (message.method) {
+			case "Runtime.consoleAPICalled":
+				logConsoleMessage(message.params);
+
+				break;
+			case "Runtime.exceptionThrown": {
+				const stack = getSourceMappedStack(message.params.exceptionDetails);
+				logger.error(message.params.exceptionDetails.text, stack);
+				break;
+			}
+			default:
+				assertNever(message);
+		}
+	}
+	async onInspectorProxyWorkerRequest(
+		message: InspectorProxyWorkerOutgoingRequestBody
+	) {
+		switch (message.type) {
+			case "runtime-websocket-error":
+				logger.error(message.error);
+
+				break;
+			case "error":
+				this.emitErrorEvent("Error inside InspectorProxyWorker", message.error);
+
+				break;
+			case "debug-log":
+				logger.debug("[InspectorProxyWorker]", ...message.args);
+
+				break;
+			case "load-network-resource": {
+				assert(this.latestConfig !== undefined);
+				assert(this.latestBundle !== undefined);
+
+				let maybeContents: string | undefined;
+				if (message.url.startsWith("wrangler-file:")) {
+					maybeContents = maybeHandleNetworkLoadResource(
+						message.url.replace("wrangler-file:", "file:"),
+						this.latestBundle,
+						this.latestBundle.sourceMapMetadata?.tmpDir
+					);
```

Comment:
This internally calls bundle-allowed-paths's `isAllowSourcePath` and `isAllowedSourceMapPath` which protect against arbitrary filesystem access by only allowing to read files referenced by the `latestBundle`